pkg/gohai/platform/platform_linux.go

```diff
@@ -5,16 +5,83 @@
 
 package platform
 
-import "strings"
-
-var unameOptions = []string{"-s", "-n", "-r", "-m", "-p", "-i", "-o"}
-
-func updateArchInfo(archInfo map[string]string, values []string) {
-	archInfo["kernel_name"] = values[0]
-	archInfo["hostname"] = values[1]
-	archInfo["kernel_release"] = values[2]
-	archInfo["machine"] = values[3]
-	archInfo["processor"] = values[4]
-	archInfo["hardware_platform"] = values[5]
-	archInfo["os"] = strings.Trim(values[6], "\n")
+import (
+	"bufio"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/gohai/utils"
+	"golang.org/x/sys/unix"
+)
+
+// getUnameOS returns the same value as `uname -o`
+//
+// uname uses preprocessor macros to determine the printable name of the operating system,
+// cf https://github.com/coreutils/gnulib/blob/master/m4/host-os.m4
+//
+// on every linux device it prints "GNU/Linux", probably because uname is from GNU coreutils,
+// so if you are using it your OS is considered to be GNU
+func getUnameOS() string {
+	// we might want to return different values depending on the actual OS (not all Linux are GNU)
+	// but for now it's good enough
+	return "GNU/Linux"
+}
+
+// from the 80_fedora_sysinfo.patch uname patch
+func procIsAthlon(reader io.Reader) bool {
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		text := scanner.Text()
+		if strings.HasPrefix(text, "vendor_id") {
+			if strings.Contains(text, "AuthenticAMD") {
+				return true
+			}
+			break
+		}
+	}
+	return false
+}
+
+// getUnameProcessor returns the same value as `uname -p`
+//
+// note that each OS applies a set of patches on the base coreutils source, in particular the 80_fedora_sysinfo.patch
+// which makes uname work on Linux, so you have to look at the patched code, eg. for ubuntu jammy:
+// https://launchpad.net/ubuntu/+archive/primary/+sourcefiles/coreutils/8.32-4.1ubuntu1/coreutils_8.32-4.1ubuntu1.debian.tar.xz
+func getUnameProcessor(machine string) string {
+	if machine == "i686" {
+		file, err := os.Open("/proc/cpuinfo")
+		if err == nil {
+			defer file.Close()
+			if procIsAthlon(file) {
+				machine = "athlon"
+			}
+		}
+	}
+
+	return machine
+}
+
+// getUnameProcessor returns the same value as `uname -i`
+//
+// note that each OS applies a set of patches on the base coreutils source, in particular the 80_fedora_sysinfo.patch
+// which makes uname work on Linux, so you have to look at the patched code, eg. for ubuntu jammy:
+// https://launchpad.net/ubuntu/+archive/primary/+sourcefiles/coreutils/8.32-4.1ubuntu1/coreutils_8.32-4.1ubuntu1.debian.tar.xz
+func getUnameHardwarePlatform(machine string) string {
```

Comment:
I am wondering if we should just keep `uname` binary when there is a no available function for a metric. As a user, I expect to have the same value from gohai as from `uname`. Also if the implementation of uname is updated, then this code also has to be updated.

Reply:
I don't think not having the exact same values as `uname` is always an issue, for example here where see that we actually have "better" values on Debian.

I'd say not depending on an external binary is important: we won't have issues when running on a platform where `uname` is not installed, we know precisely where each value comes from, and we can even improve it if we see something is missing (eg. the OS name which is considered by `uname` to always be "GNU/Linux" for every Linux, we could instead parse `/etc/os-release`).

However it's true that it moves the maintaining burden from the distro to us.

Reply:
> As a user, I expect to have the same value from gohai as from uname

QQ: is gohai supposed to provide the exact same values as GNU utilities ? Or do we just want to provide accurate information  to users ?

> Also if the implementation of uname is updated, then this code also has to be updated.

True, but I don't believe the code is moving a lot and the logic does not seem extra complicated. But yes in the current we're decoupling gohai output with uname version (and coupling gohai output with agent version) which is not really convenient imho

Reply:
I agree with @pgimalac here. The goal is to report the most accurate information to users not what uname does. We use to rely on `uname` because it was easy, but we reach the limit this solution I think.

But consistency is important, we should make sure that the new implementation doesn't introduce regression in the backend.